### PR TITLE
Feature/indeterminate checkbox

### PR DESCRIPTION
**Pull request by VilleMiekkoja**
## Description

Implements indeterminate checkbox to hds-react and hds-core

## Related Issue

https://helsinkisolutionoffice.atlassian.net/browse/HDS-1207

## How Has This Been Tested?

- Locally on developers machine
- Automated visual regression tests (loki)

## Screenshots (if appropriate):

image

👉 Demo: https://city-of-helsinki.github.io/hds-demo/indeterminate-checkbox/?path=/story/components-checkbox--indeterminate

👉 Core Demo: https://city-of-helsinki.github.io/hds-demo/indeterminate-checkbox-core/?path=/story/components-checkbox--indeterminate

👉 Checkbox Group demo: https://city-of-helsinki.github.io/hds-demo/indeterminate-checkbox/?path=/story/components-checkbox--group-with-parent

### CODE REVIEW

File: packages/react/src/components/checkbox/Checkbox.stories.tsx [hunk 1]
```diff
@@ -20,15 +20,19 @@ export default {
...
 };
 
-export const Default = () => <Checkbox id="checkbox" label="Label" />;
+export const Default = () => <Checkbox id="default" label="Label" />;
 
-export const Selected = () => <Checkbox id="checkbox2" label="Label" checked />;
+export const Selected = () => <Checkbox id="selected" label="Label" checked />;
 
-export const Disabled = () => <Checkbox id="checkbox3" label="Label" disabled />;
+export const Indeterminated = () => (
```
Comment (ronijaakkola) [inline packages/react/src/components/checkbox/Checkbox.stories.tsx:27]: A better name for the story would be "Intermediate".

File: packages/core/src/components/checkbox/checkbox.stories.js [hunk 2]
```diff
@@ -20,25 +21,39 @@ export const Selected = () => `
     </div>
 `;
 
-export const Disabled = () => `
+export const Indeterminated = () => {
```
Comment (ronijaakkola) [inline packages/core/src/components/checkbox/checkbox.stories.js:24]: A better name for the story would be "Intermediate".

**Review by ronijaakkola — COMMENTED** [review]
Visually and functionally looks good. Works as expected on VoiceOver and NVDA. 

If interested, VO announces this state "mixed" and NVDA "half-checked".

Is there a demo for the Core version? We will need that for the accessibility audit eventually anyway.

**Comment by ronijaakkola** [comment]
Also, needs documentation.

**Comment by VilleMiekkoja** [comment]
> Visually and functionally looks good. Works as expected on VoiceOver and NVDA.
> 
> If interested, VO announces this state "mixed" and NVDA "half-checked".
> 
> Is there a demo for the Core version? We will need that for the accessibility audit eventually anyway.

I can upload a demo for the core version:

- [ ] Upload demo for core version

**Comment by VilleMiekkoja** [comment]
> Also, needs documentation. In the documentation site some kind of actual example would be nice. Are we able to do something like we had in the Abstract example?

Good idea. I think we can make similar checkbox group:

- [ ] Create checkbox group with indermediate checkbox

**Review by minevala — COMMENTED** [review]
Looks good 👍  I think it is ready. The stories could be real-life checkbox group examples.

File: packages/react/src/components/checkbox/Checkbox.stories.tsx [hunk 3]
```diff
@@ -106,8 +122,16 @@ export const Playground = (args) => {
...
+      <Checkbox
+        id="indeterminate-playground"
+        label="Option 4"
+        style={styles}
+        checked={indeterminateState === 'checked'}
+        indeterminate={indeterminateState === 'indeterminate'}
+        onChange={handleIndeterminateChange}
+      />
+      <Checkbox id="checkbox7" label="Option 5" style={styles} disabled />
+      <Checkbox id="checkbox8" label="Option 6" style={styles} checked disabled />
```
[the remark this answers is not in the document]
Comment (VilleMiekkoja) [reply packages/react/src/components/checkbox/Checkbox.stories.tsx:268]: - [ ] Create a more realistic checkbox group story

**Review by minevala — APPROVED** [review]
Nice addition 👍

File: site/docs/components/checkbox.mdx [hunk 4]
```diff
@@ -158,6 +158,152 @@ Related checkboxes can be grouped together with the Selection group component. S
 </SelectionGroup>
 ```
 
+### Indeterminate
+Checkboxes also have a "mixed" state that is called an indeterminate state. This state is meant to be between the checked and unchecked states. While not very common, the usual use case for the indeterminate state is nested checkboxes. The parent checkbox can be in the indeterminate state when only part of the children checkboxes are checked.
+
+Note! You will always have to use JavaScript to set a checkbox to the indeterminate state. Currently, it cannot be done through HTML.
```
[the remark this answers is not in the document]
Comment (ronijaakkola) [reply site/docs/components/checkbox.mdx]: - [x] Added a note about indeterminate property to the Checkbox documentation

**Review by ronijaakkola — APPROVED** [review]

**Pull request merged**